Persist remember me token

hedgehog1833 opened the pull request:
Look at this beatiful PR :-)
The remember me token is now stored in a database table. The bug with the redirections is fixed. For a better user experience I added a redirect from the index page for already logged in users.
If you desperately want something to complain about: the two tests in the IndexControllerIT only test two of three endpoints. For some reason the empty_index endpoint gives a redirect in this test configuration. But after one day of building this I don't care anymore, espeacially as it works correctly in the Browser/Postman.

hedgehog1833 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Store remember me token in database](https://trello.com/c/XGhQPzn7/63-store-remember-me-token-in-database)

DanielW1987 commented on src/main/java/com/metalr2/security/SecurityConfig.java:
```diff
@@ -49,7 +55,7 @@ protected void configure(HttpSecurity http) throws Exception {
         .antMatchers(Endpoints.AntPattern.ADMIN).hasRole(UserRole.ROLE_ADMINISTRATOR.getName())
         .antMatchers(Endpoints.AntPattern.RESOURCES).permitAll()
-        .antMatchers(Endpoints.AntPattern.AUTH_PAGES).anonymous()
+        .antMatchers(Endpoints.AntPattern.AUTH_PAGES).permitAll()
```
Can a user who has been logged in now access the login page again or the register page?

DanielW1987 commented on src/main/java/com/metalr2/security/SecurityConfig.java:
```diff
@@ -59,8 +65,9 @@ protected void configure(HttpSecurity http) throws Exception {
-        .key(REMEMBER_ME_SECRET)
         .tokenValiditySeconds((int) Duration.ofDays(14).toSeconds())
+        .rememberMeServices(persistentTokenBasedRememberMeServices())
+        .key(REMEMBER_ME_SECRET)
```
you can do it as follows:

```java
.rememberMe()
  .key(REMEMBER_ME_SECRET)
  .tokenRepository(persistentTokenRepository)
  .userDetailsService(userDetailsService)
```

After that you don't need the bean `persistentTokenBasedRememberMeServices`

DanielW1987 commented on src/main/java/com/metalr2/security/SecurityConfig.java:
```diff
@@ -78,4 +85,17 @@ protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+    jdbcTokenRepository.setDataSource(dataSource);
+    return jdbcTokenRepository;
+  }
+
```
You only need the following:

```java
@Bean
public PersistentTokenRespository persistentTokenRespository() {
  JdbcTokenRepositoryImpl jdbcTokenRepository = new JdbcTokenRepositoryImpl();
  jdbcTokenRepository.setCreateTableOnStartup(false);  // is this Default?
  jdbcTokenRepository.setDataSource(dataSource);

  return jdbcTokenRepository;
}
```

DanielW1987 commented on src/main/java/com/metalr2/security/SecurityConfig.java:
```diff
@@ -78,4 +85,17 @@ protected void configure(AuthenticationManagerBuilder auth) throws Exception {
 
+  @Bean
+  public PersistentTokenBasedRememberMeServices persistentTokenBasedRememberMeServices() {
+    return new PersistentTokenBasedRememberMeServices(REMEMBER_ME_SECRET, userService, jdbcTokenRepository());
```
You don't need separate bean `persistentTokenBasedRememberMeServices `

DanielW1987 commented on src/main/java/com/metalr2/web/controller/mvc/IndexController.java:
```diff
@@ -10,8 +11,10 @@
+  public ModelAndView showIndex(Authentication authentication) {
+    if (authentication != null && authentication.isAuthenticated()) {
+      return new ModelAndView("redirect:" + Endpoints.Frontend.HOME);
+    }
```
I thought that you can define a corresponding setting in the `SecurityConfig` that, for example, always redirects to page XY by default after login.

Also: Suppose I call page `/my-artists`, but I am not logged in. I would expect to be redirected to the login page, log in and then be redirected back to the `/my-artists` page. But this does not happen. But I am of the opinion that this has worked last year. 

These should all be settings that you can make in the `SecurityConfig `.

DanielW1987 commented on src/main/java/com/metalr2/config/jpa/DatabaseConfiguration.java:
```diff
@@ -32,4 +44,13 @@
+    dataSource.setUsername(environment.getProperty("spring.datasource.username"));
+    dataSource.setPassword(environment.getProperty("spring.datasource.password"));
+    return dataSource;
+  }
```
Have you tried to autowire the `dataSource` Bean without explicit declaration?

hedgehog1833 replied:
Lol, that just f***ing works xD

DanielW1987 commented:
@hedgehog1833 
Please see my last commit. I think it's done now.

hedgehog1833 commented:
I implemented the interceptor and added some tests. I removed the check for the endpoint from the preHandle method as the interceptor is configured to run only on certain endpoints. This is tested in an integration test. preHandle functionality is tested in a unit test.

DanielW1987 commented:
In the tests, I have introduced the two classes `BaseWebMvcTest` and `BaseWebMvcTestWithSecurity`. This saves the whole boilerplate code around mocking. Please have a final look before merging.
